framework/math/SpatialDiscretization/CellMappings/LagrangeBaseMapping.h

```diff
@@ -0,0 +1,116 @@
+  MakeVolumetricQuadraturePointData() const override;
+
+  finite_element::SurfaceQuadraturePointData
+  MakeSurfaceQuadraturePointData(size_t face_index) const override;
+
+protected:
+  friend class LagrangeWorldXYZToNaturalMappingHelper;
+  typedef chi_mesh::Vector3 Vec3;
+  typedef std::vector<double> VecDbl;
+  typedef std::vector<VecDbl> MatDbl;
+
+  LagrangeBaseMapping(const chi_mesh::MeshContinuum& grid,
+                      const chi_mesh::Cell& cell,
+                      size_t num_nodes,
+                      std::vector<std::vector<int>> face_node_mappings,
+                      const Quadrature& volume_quadrature,
+                      const Quadrature& surface_quadrature);
+
+  /**Function using newton iteration to convert a world_xyz coordinate to
+   * a quadrature point coordinate.*/
+  Vec3 MapWorldXYZToNaturalXYZ(const Vec3& world_xyz) const;
+
+  /**Used in the constructor this method simply collects the cell vertices.*/
+  static std::vector<chi_mesh::Vector3>
+  GetVertexLocations(const chi_mesh::MeshContinuum& grid,
+                     const chi_mesh::Cell& cell);
+
+  /**Used in the constructor this method maps face node ids to cell ids.*/
+  static std::vector<std::vector<int>>
+  MakeFaceNodeMapping(const chi_mesh::Cell& cell);
+
+  /**Reference element shape function evaluation. Meant to be overridden by
+   * a specific element mapping.*/
+  virtual double RefShape(uint32_t i, const Vec3& qpoint) const = 0;
+
+  /**Reference element shape function gradient evaluation. Meant to be
+   * overridden by a specific element mapping.*/
+  virtual Vec3 RefGradShape(uint32_t i, const Vec3& qpoint) const = 0;
+
+  /**Reference element Jacobian evaluation. Meant to be overridden by
+   * a specific element mapping.*/
+  virtual MatDbl RefJacobian(const Vec3& qpoint) const = 0;
+
+  /**Computes the determinant of the surface Jacobian and the surface normal
+   * at the face quadrature point.*/
+  virtual std::pair<double, Vec3>
+  RefFaceJacobianDeterminantAndNormal(size_t face_index,
+                                      const Vec3& qpoint_face) const;
+
+  /**Converts a face quadrature point to a reference element position allowing
+   * the reuse of the reference element shape functions. Meant to be overridden
+   * by a specific element mapping.*/
+  virtual Vec3 FaceToElementQPointConversion(size_t face_index,
+                                             const Vec3& qpoint_face) const = 0;
+
+  /**Overrideable method to return the appropriate face quadrature. Default
+   * returns surface_quadrature. Some mappings might change quadrature based
+   * on the face index.*/
+  virtual const Quadrature& GetSurfaceQuadrature(size_t face_index) const;
+
+  const Quadrature& volume_quadrature_;
+  const Quadrature& surface_quadrature_;
+};
+
+/**Helper class to apply Newton's method to map world coordinates to
+ * reference element coordinates.*/
+class LagrangeWorldXYZToNaturalMappingHelper : public NonLinearFunction
```

Comment:
I think I am going to stick with World and Natural here to match the textbook I based it off. There are also tons of google-able stuff popping up when I search for something like `quad natural coordinates` as opposed to `quad reference coordinates`. Deal.ii uses RealToUnitCell but I don't really like the "real" part.

I did remove the Lagrange part here to reduce confusion however when we have other methods doing something similar we will have segregate it again.

Reply:
World and Natural are fine as long as it is documented. I would recommend `NaturalXYZ` in the class name as well. I don't know what other methods of mapping coordinates you plan to put in (or if it will be in this class or not), but we'll cross that bridge when we get to it.